#! /bin/sh /usr/share/dpatch/dpatch-run
## MSA-09-0009.dpatch by Dan Poltawski <talktodan@gmail.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Tex fitler allows for files to be disclosed

@DPATCH@

diff --git a/filter/tex/filter.php b/filter/tex/filter.php
index e8b0709..1c6d348 100644
--- a/filter/tex/filter.php
+++ b/filter/tex/filter.php
@@ -118,6 +118,16 @@ function tex_filter ($courseid, $text) {
         $text = str_replace($matches[0][$i],$replacement,$text);
     }
 
+    // TeX blacklist. MDL-18552
+    $tex_blacklist = array(
+        'include','def','command','loop','repeat','open','toks','output',
+        'input','catcode','name','^^',
+        '\every','\errhelp','\errorstopmode','\scrollmode','\nonstopmode',
+        '\batchmode','\read','\write','csname','\newhelp','\uppercase',
+        '\lowercase','\relax','\aftergroup',
+        '\afterassignment','\expandafter','\noexpand','\special'
+    );
+
     // <tex> TeX expression </tex>
     // or $$ TeX expression $$
     // or \[ TeX expression \]          // original tag of MathType and TeXaide (dlnsk)
@@ -138,6 +148,19 @@ function tex_filter ($courseid, $text) {
           $align = "text-top";
           $texexp = preg_replace('/^align=top /','',$texexp);
         }
+    /// Check $texexp against blacklist (whitelisting could be more complete but also harder to maintain). MDL-18552
+        $invalidcommands = array();
+        foreach($tex_blacklist as $command) {
+            if (stristr($texexp, $command)) { /// Found invalid command. Annotate.
+                $invalidcommands[] = $command;
+            }
+        }
+        if (!empty($invalidcommands)) { /// Invalid commands found. Output error and continue with next TeX element
+            $invalidstr = get_string('invalidtexcommand', 'error', implode(', ', $invalidcommands));
+            $text = str_replace( $matches[0][$i], $invalidstr, $text);
+            continue;
+        }
+    /// Everything is ok, let's process the expression
         $md5 = md5($texexp);
         if (! $texcache = get_record("cache_filters","filter","tex", "md5key", $md5)) {
             $texcache->filter = 'tex';
diff --git a/filter/tex/texdebug.php b/filter/tex/texdebug.php
index a2b95d9..272b904 100644
--- a/filter/tex/texdebug.php
+++ b/filter/tex/texdebug.php
@@ -3,8 +3,6 @@
       // If not, it obtains the corresponding TeX expression from the cache_tex db table
       // and uses mimeTeX to create the image file
 
-    $nomoodlecookie = true;     // Because it interferes with caching
-
     require_once("../../config.php");
 
     if (empty($CFG->textfilters)) {
@@ -19,6 +17,9 @@
     $CFG->texfilterdir = "filter/tex";
     $CFG->teximagedir = "filter/tex";
 
+    require_login();
+    require_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM), $USER->id); /// Required cap to run this. MDL-18552
+
     $query = urldecode($_SERVER['QUERY_STRING']);
     error_reporting(E_ALL);
 
